commit 5a2b7c25b68f2c955356640041e4c7ed72416d4e
Author: Pere Diaz Bou <pdiazbou@redhat.com>
Date:   Fri Nov 11 10:43:01 2022 +0100

    mgr/prometheus: expose daemon health metrics
    
    Until now daemon health metrics were stored without being used. One of
    the most helpful metrics there is SLOW_OPS with respect to OSDs and MONs
    which this commit tries to expose to bring fine grained metrics to find
    troublesome OSDs instead of having a lone healthcheck of slow ops in the
    whole cluster.
    
    Signed-off-by: Pere Diaz Bou <pdiazbou@redhat.com>

diff --git a/src/mgr/ActivePyModules.cc b/src/mgr/ActivePyModules.cc
index 58c3d9ee4d6..8c070dd8cbf 100644
--- a/src/mgr/ActivePyModules.cc
+++ b/src/mgr/ActivePyModules.cc
@@ -1515,3 +1515,26 @@ void ActivePyModules::unregister_client(std::string_view name, std::string addrs
   dout(7) << "unregistering msgr client handle " << addrv << dendl;
   py_module_registry.unregister_client(name, addrv);
 }
+
+PyObject* ActivePyModules::get_daemon_health_metrics()
+{
+  without_gil_t no_gil;
+  return daemon_state.with_daemons_by_server([&no_gil]
+      (const std::map<std::string, DaemonStateCollection> &all) {
+      no_gil.acquire_gil();
+      PyFormatter f;
+      for (const auto &[hostname, daemon_state] : all) {
+        for (const auto &[key, state] : daemon_state) {
+          f.open_array_section(ceph::to_string(key));
+          for (const auto &metric : state->daemon_health_metrics) {
+            f.open_object_section(metric.get_type_name());
+            f.dump_int("value", metric.get_n1());
+            f.dump_string("type", metric.get_type_name());
+            f.close_section();
+          }
+          f.close_section();
+        }
+      }
+      return f.get();
+  });
+}
diff --git a/src/mgr/ActivePyModules.h b/src/mgr/ActivePyModules.h
index f283224f588..8f70b4fe05a 100644
--- a/src/mgr/ActivePyModules.h
+++ b/src/mgr/ActivePyModules.h
@@ -222,6 +222,7 @@ public:
 
   void cluster_log(const std::string &channel, clog_type prio,
     const std::string &message);
+  PyObject* get_daemon_health_metrics();
 
   bool inject_python_on() const;
   void update_cache_metrics();
